src/ripple/app/misc/detail/WorkFile.h

```diff
@@ -0,0 +1,144 @@
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+#define RIPPLE_APP_MISC_DETAIL_WORKFILE_H_INCLUDED
+
+#include <ripple/app/misc/detail/Work.h>
+#include <boost/system/error_code.hpp>
+#include <cassert>
+
+namespace ripple {
+
+namespace detail {
+
+// Work with files
+class WorkFile: public Work
+    , public std::enable_shared_from_this<WorkFile>
+{
+protected:
+    using error_code = boost::system::error_code;
+    // Override the definition in Work.h
+    using response_type = std::string;
+
+public:
+    using callback_type =
+        std::function<void(error_code const&, response_type&&)>;
+public:
+    WorkFile(
+        std::string const& path,
+        boost::asio::io_service& ios, callback_type cb);
+    ~WorkFile();
+
+    void run() override;
+
+    void cancel() override;
+
+private:
+    std::string path_;
+    callback_type cb_;
+    boost::asio::io_service& ios_;
+    boost::asio::io_service::strand strand_;
+
+};
+
+//------------------------------------------------------------------------------
+
+WorkFile::WorkFile(
+    std::string const& path,
+    boost::asio::io_service& ios, callback_type cb)
+    : path_(path)
+    , cb_(std::move(cb))
+    , ios_(ios)
+    , strand_(ios)
+{
+}
+
+WorkFile::~WorkFile()
+{
+    if (cb_)
+        cb_ (make_error_code(boost::system::errc::interrupted),
+            {});
+}
+
+void
+WorkFile::run()
+{
+    if (! strand_.running_in_this_thread())
+        return ios_.post(strand_.wrap (std::bind(
+            &WorkFile::run, shared_from_this())));
+
+    error_code ec;
+    auto fileContents = [&]()
+        -> std::string
+    {
+        using namespace boost::filesystem;
+#if _MSC_VER    // MSVC
+        {
+            auto first = path_.find_first_not_of("\\/");
+            if (first != 0)
+                path_.erase(path_.begin(), path_.begin() + first);
+
+        }
+#endif
+        path bestPath{ canonical(path_, ec) };
+        if (ec)
+        {
+            return {};
+        }
+        std::ifstream fileStream(bestPath.c_str(), std::ios::in);
+
+        if (!fileStream)
+        {
+            std::cerr << "Failed to open '" << bestPath << "'." << std::endl;
+            ec = make_error_code(boost::system::errc::no_such_file_or_directory);
+            return {};
+        }
+
+        std::string result;
+        result.assign ((std::istreambuf_iterator<char>(fileStream)),
```

Comment:
Extra pair of parens `std::istreambuf_iterator<char>(fileStream)` here.

Also, we should be able to directly initialize `result` here:

    std::string result(
        std::istreambuf_iterator<char>(fileStream),
        std::istreambuf_iterator<char>());

However, I've never been able to do that and always end up getting errors I've given up trying to decipher. Paging @HowardHinnant. Am I missing something?

Reply:
You got hit by the infamous "most vexing parse."  `std::istreambuf_iterator<char>()` gets interpreted as a function declaration (for reasons I do not understand).  But the fix is easy:

```
        std::string result(
            std::istreambuf_iterator<char>(fileStream),
            std::istreambuf_iterator<char>{});
```

The `{}` on the second arg does it.  Heck, I would splurge and use `{}` for both constructions:

```
        std::string result(
            std::istreambuf_iterator<char>{fileStream},
            std::istreambuf_iterator<char>{});
```

Reply:
Huh, interesting. I'm familiar with "most vexing parse" issue but it never occurred to me that this was what was happening. Thanks!

Reply:
Fixed, and refactored out into a separate function for reuse in the other places I've seen this method of reading a file.